reco_utils/recommender/deeprec/models/graphrec/lightgcn.py

```diff
@@ -0,0 +1,403 @@
+            self.ia_embeddings, self.neg_items
+        )
+        self.u_g_embeddings_pre = tf.nn.embedding_lookup(
+            self.weights["user_embedding"], self.users
+        )
+        self.pos_i_g_embeddings_pre = tf.nn.embedding_lookup(
+            self.weights["item_embedding"], self.pos_items
+        )
+        self.neg_i_g_embeddings_pre = tf.nn.embedding_lookup(
+            self.weights["item_embedding"], self.neg_items
+        )
+
+        self.batch_ratings = tf.matmul(
+            self.u_g_embeddings,
+            self.pos_i_g_embeddings,
+            transpose_a=False,
+            transpose_b=True,
+        )
+
+        self.mf_loss, self.emb_loss = self._create_bpr_loss(
+            self.u_g_embeddings, self.pos_i_g_embeddings, self.neg_i_g_embeddings
+        )
+        self.loss = self.mf_loss + self.emb_loss
+
+        self.opt = tf.train.AdamOptimizer(learning_rate=self.lr).minimize(self.loss)
+        self.saver = tf.train.Saver(max_to_keep=1)
+
+        gpu_options = tf.GPUOptions(allow_growth=True)
+        self.sess = tf.Session(config=tf.ConfigProto(gpu_options=gpu_options))
+        self.sess.run(tf.global_variables_initializer())
+
+    def _init_weights(self):
+        """Initialize user and item embeddings.
+
+        Returns:
+            dict: With keys "user_embedding" and "item_embedding", embeddings of all users and items.
+
+        """
+        all_weights = dict()
+        initializer = tf.contrib.layers.xavier_initializer()
+
+        all_weights["user_embedding"] = tf.Variable(
+            initializer([self.n_users, self.emb_dim]), name="user_embedding"
+        )
+        all_weights["item_embedding"] = tf.Variable(
+            initializer([self.n_items, self.emb_dim]), name="item_embedding"
+        )
+        print("Using xavier initialization.")
+
+        return all_weights
+
+    def _create_lightgcn_embed(self):
+        """Calculate the average embeddings of users and items after every layer of the model.
+
+        Returns:
+            tf.tensor: average user embeddings
+            tf.tensor: average item embeddings
+
+        """
+        A_hat = self._convert_sp_mat_to_sp_tensor(self.norm_adj)
+
+        ego_embeddings = tf.concat(
+            [self.weights["user_embedding"], self.weights["item_embedding"]], axis=0
+        )
+        all_embeddings = [ego_embeddings]
+
+        for k in range(0, self.n_layers):
+            ego_embeddings = tf.sparse_tensor_dense_matmul(A_hat, ego_embeddings)
+            all_embeddings += [ego_embeddings]
+
+        all_embeddings = tf.stack(all_embeddings, 1)
+        all_embeddings = tf.reduce_mean(all_embeddings, axis=1, keepdims=False)
+        u_g_embeddings, i_g_embeddings = tf.split(
+            all_embeddings, [self.n_users, self.n_items], 0
+        )
+        return u_g_embeddings, i_g_embeddings
+
+    def _create_bpr_loss(self, users, pos_items, neg_items):
+        """Calculate BPR loss.
+
```

Comment:
explain the arguments here, just as how you did for other functions:
        Args:
            user_ids (np.array): Users to test.
            remove_seen (bool): Flag to remove items seen in training from recommendation.